### ISSUE TITLE
Memory issue when using virtuous-rub

### ISSUE THREAD

**Issue by js-john** [utmapp/UTM#2720]
**Describe the issue**
A clear and concise description of what the issue is. **BEFORE SUBMITTING YOUR ISSUE, PLEASE LOOK AT THE PINNED ISSUES AND USE THE SEARCH FUNCTION TO MAKE SURE IT IS NOT ALREADY REPORTED. ALWAYS COMMENT ON AN EXISTING ISSUE INSTEAD OF MAKING A NEW ONE.**

**Configuration (required)**
* UTM Version: 
* OS Version: 
* Device Model: 
* Is it jailbroken (name jailbreak used)? 
* How did you install UTM? 

**Crash log**
If the app crashed, you need a crash log. To get your crash log, open the Settings app and browse to `Privacy -> Analytics & Improvements -> Analytics Data` and find the latest entry for UTM. You should export the text and attach it here.

**Debug log**
For all issues, _including_ crashes, you should attach a debug log. Open UTM, and open the settings for the VM you wish to launch. Near the bottom of the configuration options (or at the top of the `QEMU` page) is `Debug Log`. Turn it on and save the VM. After you experience the issue, quit the VM and re-launch UTM. Open the VM settings again and select `Export Log...` and attach it here.

**Upload VM**
(Optional) If possible, upload the `config.plist` inside the `.utm`. If you do not have this, you can upload the entire `.utm` but note this contains your personal data. Since GitHub has an attachment size limit, you may want to upload to another service such as Google Drive. Link it here.

**Comment by js-john** [comment]
![2B87E28C-8A19-4C0A-BED3-63CFE5CB5AFD](https://user-images.githubusercontent.com/9637998/128751028-8e7d82e7-5ff0-408d-a25d-d38980ed026c.jpeg)

**Comment by conath** [comment]
Thanks for this. A couple of questions: is this new behavior in the 2.2.0 beta? Did you change the JIT cache setting for this VM?

Anyway, this does seem off, as the estimate in the Config code comes to just 3200MB used when setting RAM to 2GB and JIT cache at default. https://github.com/utmapp/UTM/blob/d915297066aaebab5e933bbb857e4456af0623ab/Platform/Shared/VMConfigSystemView.swift#L88

**Comment by js-john** [comment]
Yes, this issue only appears in 2.2.0 Beta. I leave the JIT catch setting as default.

**Comment by osy** [comment]
I think this is inevitable with OpenGL because the video memory used is counted separately by iOS.

**Comment by conath** [comment]
I have what must be a memory leak when running 3D OpenGL on macOS: 
<img width="1440" alt="Screen Shot 2021-08-11 at 01 36 36" src="https://user-images.githubusercontent.com/12073163/128948432-d4048169-3919-482b-8a3a-d64dcc7118f0.png">
<img width="943" alt="Screen Shot 2021-08-11 at 01 37 14" src="https://user-images.githubusercontent.com/12073163/128948442-39bc2f58-087b-40a1-96b0-f919afc6533d.png">

**Comment by coreycusick** [comment]
Ah. So not just me having issues with memory. Crashes when 3D acceleration is used at all with an out of memory error on M1 iPad Pro 11 inch.!Crashes as soon as any OpenGL application is used

**Comment by lulor** [comment]
> I have what must be a memory leak when running 3D OpenGL on macOS:
> <img alt="Screen Shot 2021-08-11 at 01 36 36" width="1440" src="https://user-images.githubusercontent.com/12073163/128948432-d4048169-3919-482b-8a3a-d64dcc7118f0.png">
> <img alt="Screen Shot 2021-08-11 at 01 37 14" width="943" src="https://user-images.githubusercontent.com/12073163/128948442-39bc2f58-087b-40a1-96b0-f919afc6533d.png">

I have encountered the same issue with UTM 2.2.0 Beta on an M1 Mac

**Comment by deja-geek** [comment]
I've encountered the same memory leak on UTM 2.2.0 Beta on an M1 Mac using both the virtio-ramfb and virtio-gpu-gl-pci adapter.  VM running is Ubuntu 20.04 LTS running GNOME

**Comment by mdillavou** [comment]
I tried the 2.2.0 Beta on an M1 Mac and had to revert back to 2.1.2 as I kept running out of memory. I also had switched to `virtio-gpu-gl-pci` adapter, which gave me rendering artifacts, so I then switched to `virtio-ramfb`

**Comment by 1640675651** [comment]
Same problem on my 4GB iPad pro. When I assign 2048MB RAM to the VM, UTM says it will approximately take ~2700MB of RAM, but it got killed and Jetsam event log says it takes 3GB of RAM.

**Comment by osy** [comment]
I've spent the weekend trying to root cause the issue and I'm still stuck. For some reason, textures allocated in ANGLE are not being deallocated properly. Using the Metal GPU debugger, we find that a lot of the textures are left in memory even though they look like stale data. I'm not sure exactly why the resources are not being freed. It may be an issue with the `autoreleasepool` but I can't find anything there.

https://github.com/utmapp/angle/commit/ea4666413921445dd597da6063e2681e4874eefb is a workaround by marking the memory as purgeable when it's being deallocated by ANGLE. It fixes the memory leak issue but is not a good solution because you get random GPU artifacts sometimes as it tries to use the purged memory. In fact, if you set it to force deallocate with `MTLPurgeableStateEmpty` then a lot of graphics won't render. So there is some race between ANGLE being done with a texture and it being rendered on screen.

@akihikodaki Have you had any issues with memory leak using ES profile with ANGLE on your QEMU implementation?

**Comment by akihikodaki** [comment]
@osy No, I'm not aware of memory leak. I assume you use Metal backend, but it is not ready and should be used only for limited inputs or testing.
